### PR TITLE
kubeadm: Move all node bootstrap token related code in one phase package

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Part of the phases refactoring.
Moves everything Node Bootstrap Token-related into its own package.
In the future there will be a `phases/bootstraptoken/master` pkg as well.
The generic bootstrap token client functions should be moved to client go eventually https://github.com/kubernetes/client-go/issues/114

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
I'll yet add the CLI interface for this tomorrow.
Not sure if this compiles currently, but I'm uploading this now for initial review.

**Release note**:

```release-note
NONE
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @mattmoyer

### CODE REVIEW

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1beta1"
+	bootstrapapi "k8s.io/kubernetes/pkg/bootstrap/api"
+	"k8s.io/kubernetes/pkg/util/version"
+)
+
+const (
+	// nodeBootstrapperClusterRoleName defines the name of the auto-bootstrapped ClusterRole for letting someone post a CSR
+	nodeBootstrapperClusterRoleName = "system:node-bootstrapper"
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: Shouldn't this leverage system consts?

File: cmd/kubeadm/app/phases/bootstraptoken/node/clusterinfo.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		Clusters: map[string]*clientcmdapi.Cluster{
+			"": adminConfig.Clusters[adminCluster],
+		},
+	}
+	bootstrapBytes, err := clientcmd.Write(*bootstrapConfig)
+	if err != nil {
+		return err
+	}
+
+	bootstrapConfigMap := v1.ConfigMap{
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/node/clusterinfo.go]: perhaps @ericchiang &| @mattmoyer can once over this PR.  

I'm not intimately familiar with the ordering of bootstrap-signing and if this is all kosher.

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 1 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: Doesn't exist :(

File: cmd/kubeadm/app/constants/constants.go [hunk 3]
```diff
@@ -121,6 +121,9 @@ var (
 
 	// MinimumControlPlaneVersion specifies the minimum control plane version kubeadm can deploy
 	MinimumControlPlaneVersion = version.MustParseSemantic("v1.7.0")
+
+	// MinimumCSRAutoApprovalClusterRolesVersion
+	MinimumCSRAutoApprovalClusterRolesVersion = version.MustParseSemantic("v1.7.0-alpha.3")
```
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: Thanks for that catch! Should have been v1.8.0-alpha.3 indeed as the PR just merged

File: cmd/kubeadm/app/phases/bootstraptoken/node/clusterinfo.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/clusterinfo.go]: This should be just moving Go code around, but sure, it would be great if for example @mattmoyer could take a pass at this :+1:

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+// AutoApproveNodeBootstrapTokens creates RBAC rules in a way that makes Node Bootstrap Tokens' CSR auto-approved by the csrapprover controller
+func AutoApproveNodeBootstrapTokens(client clientset.Interface, k8sVersion *version.Version) error {
+	// TODO: When the v1.9 cycle starts (targeting v1.9 at HEAD) and v1.8.0 is the minimum supported version, we can remove this function as the ClusterRole will always exist
+	if k8sVersion.LessThan(constants.MinimumCSRAutoApprovalClusterRolesVersion) {
+
+		err := createClusterRoleIfNotExists(client, &rbac.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: csrAutoApprovalClusterRoleName,
```
Comment (mattmoyer) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: If I'm following correctly, this cluster role used to get created as `nodeAutoApproveBootstrap = "kubeadm:node-autoapprove-bootstrap"`, now it's `csrAutoApprovalClusterRoleName = "system:certificates.k8s.io:certificatesigningrequests:nodeclient"`. Expected?

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+				Name: bootstrapapi.BootstrapGroup,
+			},
+		},
+	})
+}
+
+// AutoApproveNodeBootstrapTokens creates RBAC rules in a way that makes Node Bootstrap Tokens' CSR auto-approved by the csrapprover controller
+func AutoApproveNodeBootstrapTokens(client clientset.Interface, k8sVersion *version.Version) error {
+	// TODO: When the v1.9 cycle starts (targeting v1.9 at HEAD) and v1.8.0 is the minimum supported version, we can remove this function as the ClusterRole will always exist
+	if k8sVersion.LessThan(constants.MinimumCSRAutoApprovalClusterRolesVersion) {
```
Comment (mattmoyer) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: This is new code here, right? Not just copied around?

**Review by mattmoyer — COMMENTED** [review]
Left a couple of questions, but otherwise this seems like a nice refactor.

File: cmd/kubeadm/app/cmd/init.go [hunk 6]
```diff
@@ -221,15 +222,18 @@ func (i *Init) Validate(cmd *cobra.Command) error {
...
 
-	// PHASE 1: Generate certificates
-	err := cmdphases.CreatePKIAssets(i.cfg)
+	k8sVersion, err := version.ParseSemantic(i.cfg.KubernetesVersion)
 	if err != nil {
+		return fmt.Errorf("couldn't parse kubernetes version %q: %v", i.cfg.KubernetesVersion, err)
+	}
+
+	// PHASE 1: Generate certificates
+	if err := cmdphases.CreatePKIAssets(i.cfg); err != nil {
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: please use a var for err vs. repeated implicit declarations.

File: cmd/kubeadm/app/constants/constants.go [hunk 7]
```diff
@@ -121,6 +121,9 @@ var (
 
 	// MinimumControlPlaneVersion specifies the minimum control plane version kubeadm can deploy
 	MinimumControlPlaneVersion = version.MustParseSemantic("v1.7.0")
+
+	// MinimumCSRAutoApprovalClusterRolesVersion
+	MinimumCSRAutoApprovalClusterRolesVersion = version.MustParseSemantic("v1.8.0-alpha.3")
```
Comment (timothysc) [inline cmd/kubeadm/app/constants/constants.go]: I really wish there was a better way to handle new features as we shouldn't be gating on it until some reasonable level of testing.  Could you add a //TODO to convert to 1.8.0 once we near release.

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	bootstrapConfigMap := v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: bootstrapapi.ConfigMapClusterInfo},
+		Data: map[string]string{
+			bootstrapapi.KubeConfigKey: string(bootstrapBytes),
+		},
+	}
+
+	if _, err := client.CoreV1().ConfigMaps(metav1.NamespacePublic).Create(&bootstrapConfigMap); err != nil {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go]: So this is slightly inconsistent usage of implicit vars here.  Another use a var seems more appropriate.

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	bootstrapapi "k8s.io/kubernetes/pkg/bootstrap/api"
+)
+
+const (
+	// BootstrapSignerClusterRoleName sets the name for the ClusterRole that allows access to ConfigMaps in the kube-public ns
+	BootstrapSignerClusterRoleName = "kubeadm:bootstrap-signer-clusterinfo"
+)
+
+// SetupPublicClusterInfoConfigMap creates the public cluster-info ConfigMap and attached RBAC rules (if it doesn't already exist)
+func SetupPublicClusterInfoConfigMap(client clientset.Interface, file string) error {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go]: This is the same code that I wanted one of the auth folks to look at, did they verify?

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go (hunk 8 above)
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go]: It doesn't matter much, isn't the logic here kind of inverted?  You always try to create and pass if it exists, vs. checking if it exists 1st and exit early.

Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go]: Or copy the update logic like the code in the next f(n).

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 1 above)
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: could you file an issue on this.

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TODO: We should invent a dynamic mechanism for this using the dynamic client instead of hard-coding these functions per-type
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: Yeah templates/generic would be "super" ;-)

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/cmd/init.go (hunk 6 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Why? I think that `if err := foo(); err != nil {` is the cleanest way of doing things when `foo()` only returns one error.

File: cmd/kubeadm/app/constants/constants.go (hunk 7 above)
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: Added such a TODO

Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: Agree that this will be bumped higher when we can

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go (hunk 8 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo.go]: The logic was ok, but I made it more obvious with the latest rework

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 10 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: It is doable using the dynamic client (`k8s.io/client-go/dynamic`) and I have a rough idea about how to do it, but it's out of scope and too risky for v1.8

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 5 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: basically the only new thing; should be gated behind that feature gate.

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 4 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: Yes, you're following correctly and it is expected.
This ClusterRole changed name when becoming "official" (_sigh_): https://github.com/kubernetes/kubernetes/issues/48191

The most important thing is that the binding has the same name as earlier, so it will be updated to use the right ClusterRole when upgrading.

Thanks for your detailed observation, can't merge something shitty here :wink:

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 1 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: https://github.com/kubernetes/kubernetes/issues/50315

File: cmd/kubeadm/app/cmd/init.go [hunk 11]
```diff
@@ -249,31 +254,32 @@ func (i *Init) Run(out io.Writer) error {
...
+	// Create the default node bootstrap token
 	tokenDescription := "The default bootstrap token generated by 'kubeadm init'."
-	if err := tokenphase.UpdateOrCreateToken(client, i.cfg.Token, false, i.cfg.TokenTTL, kubeadmconstants.DefaultTokenUsages, tokenDescription); err != nil {
+	if err := nodebootstraptokenphase.UpdateOrCreateToken(client, i.cfg.Token, false, i.cfg.TokenTTL, kubeadmconstants.DefaultTokenUsages, tokenDescription); err != nil {
 		return err
 	}
-
-	if err := tokenphase.CreateBootstrapConfigMapIfNotExists(client, adminKubeConfigPath); err != nil {
+	// Create the cluster-info ConfigMap with the associated RBAC rules
+	if err := nodebootstraptokenphase.SetupPublicClusterInfoConfigMap(client, adminKubeConfigPath); err != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Fixed.

**Comment by luxas** [comment]
@timothysc @mattmoyer Thank you for the review. I tried to make things even more obvious in this pass.
This combined with https://github.com/kubernetes/kubernetes/pull/50214 will effectively kill the monolith `phases/apiconfig/clusterroles.go` file (and package), which will be great.

Instead of doing all kinds of different RBAC magic in one place (all-or-nothing), these are now in their own logical package and at the same time the associated object is.

I also added a lightweight CLI extension to this just for invoking the method. If you feel that that's overkill in this PR but you are happy with the rest of the refactor, I'll just rip it out, let this merge and send a new PR for the CLI commands. Whatever you like.

File: cmd/kubeadm/app/cmd/init.go (hunk 6 above)
Comment (mattmoyer) [inline cmd/kubeadm/app/cmd/init.go]: I am also interested to hear if there's a reason to prefer `err` scoped outside the `if`. I prefer this style if there's no need for the return value to be used outside the `if`.

File: cmd/kubeadm/app/cmd/init.go [hunk 12]
```diff
@@ -244,47 +248,54 @@ func (i *Init) Run(out io.Writer) error {
 		return err
 	}
 
+	// PHASE 4: Mark the master with the right label/taint
 	if err := markmasterphase.MarkMaster(client, i.cfg.NodeName); err != nil {
 		return err
 	}
 
-	// PHASE 4: Set up the bootstrap tokens
+	// PHASE 5: Set up the node bootstrap tokens
```
Comment (mattmoyer) [inline cmd/kubeadm/app/cmd/init.go]: Minor nit: it might be easier to maintain these comments if we dropped the numbering.

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go [hunk 13]
```diff
@@ -120,7 +83,7 @@ func TestCreateBootstrapConfigMapIfNotExists(t *testing.T) {
 			})
 		}
 
-		err = CreateBootstrapConfigMapIfNotExists(client, file.Name())
+		err = createBootstrapConfigMapIfNotExists(client, file.Name())
```
Comment (mattmoyer) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go]: Is this a typo? It doesn't look like the name of this function changed.

File: cmd/kubeadm/app/cmd/init.go (hunk 6 above)
Comment (jbeda) [inline cmd/kubeadm/app/cmd/init.go]: Much of the rest of the file uses this style so I think that it is best to at least be consistent here. This is a little out of step of much other go code in the k8s code base so it might be worth rationalizing at some point.

File: cmd/kubeadm/app/phases/addons/addons.go [hunk 14]
```diff
@@ -100,14 +101,8 @@ func CreateKubeProxyAddon(configMapBytes, daemonSetbytes []byte, client clientse
...
 
-	if _, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Create(kubeproxyConfigMap); err != nil {
-		if !apierrors.IsAlreadyExists(err) {
-			return fmt.Errorf("unable to create a new kube-proxy configmap: %v", err)
-		}
-
-		if _, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Update(kubeproxyConfigMap); err != nil {
-			return fmt.Errorf("unable to update the kube-proxy configmap: %v", err)
-		}
+	if err := apiclientutil.CreateConfigMapIfNotExists(client, kubeproxyConfigMap); err != nil {
```
Comment (jbeda) [inline cmd/kubeadm/app/phases/addons/addons.go]: This seems like an unrelated change.  It is probably ok to sneak this in, but it sticks out here.

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+	})
+}
+
+// AutoApproveNodeBootstrapTokens creates RBAC rules in a way that makes Node Bootstrap Tokens' CSR auto-approved by the csrapprover controller
+func AutoApproveNodeBootstrapTokens(client clientset.Interface, k8sVersion *version.Version) error {
+
+	fmt.Println("[bootstraptoken] Allowing Node Bootstrap Tokens to automatically get their unique node credential signed")
+
+	// TODO: When the v1.9 cycle starts (targeting v1.9 at HEAD) and v1.8.0 is the minimum supported version, we can remove this function as the ClusterRole will always exist
+	if k8sVersion.LessThan(constants.MinimumCSRAutoApprovalClusterRolesVersion) {
```
Comment (mattmoyer) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: Maybe there should be an extra `fmt.Println("[bootstraptoken] ...")` here mentioning whether or not this is being created?

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+const (
+	// NodeBootstrapperClusterRoleName defines the name of the auto-bootstrapped ClusterRole for letting someone post a CSR
+	// TODO: This value should be defined in an other, generic authz package instead of here
+	NodeBootstrapperClusterRoleName = "system:node-bootstrapper"
+	// NodeKubeletBootstrap defines the name of the ClusterRoleBinding that lets kubelets post CSRs
+	NodeKubeletBootstrap = "kubeadm:kubelet-bootstrap"
+
+	// CSRAutoApprovalClusterRoleName defines the name of the auto-bootstrapped ClusterRole for making the csrapprover controller auto-approve the CSR
+	// TODO: This value should be defined in an other, generic authz package instead of here
+	CSRAutoApprovalClusterRoleName = "system:certificates.k8s.io:certificatesigningrequests:nodeclient"
```
Comment (jbeda) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: This is a long role name.  Was this suggested by someone?  I wouldn't think that something under `system:...` would also need to be qualified with a domain.

File: cmd/kubeadm/app/cmd/phases/phase_test.go [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"testing"
+)
+
+func TestGetStaticPodFilepath(t *testing.T) {
```
Comment (mattmoyer) [inline cmd/kubeadm/app/cmd/phases/phase_test.go]: This looks like it should be `TestValidateExactArgNumber`.

**Review by jbeda — APPROVED** [review]
In general looks good to me.  I'll approve and let moyer give the lgtm.

File: cmd/kubeadm/app/cmd/phases/bootstraptoken.go [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+	cmd.AddCommand(NewSubCmdNodeBootstrapTokenAutoApprove(kubeConfigFile))
+
+	return cmd
+}
+
+// NewSubCmdNodeBootstrapTokenPostCSRs returns the Cobra command for running the allow-post-csrs sub-phase
+func NewSubCmdNodeBootstrapTokenPostCSRs(kubeConfigFile *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "allow-post-csrs",
+		Short: "Allow a Node Bootstrap Token post CSRs in order to get its real credential later",
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/phases/bootstraptoken.go]: "Configure RBAC to allow Node Bootstrap tokens to post CSRs in order for nodes to get long term certificate credentials."

File: cmd/kubeadm/app/cmd/phases/bootstraptoken.go [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+		},
+	}
+	return cmd
+}
+
+// NewSubCmdNodeBootstrapToken returns the Cobra command for running the allow-auto-approve sub-phase
+func NewSubCmdNodeBootstrapTokenAutoApprove(kubeConfigFile *string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "allow-auto-approve",
+		Short: "Allow the csrapprover controller automatically approve CSRs from a Node Bootstrap Token",
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/phases/bootstraptoken.go]: "Configure RBAC rules to allow..."

File: cmd/kubeadm/app/cmd/phases/bootstraptoken.go [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+		Aliases: []string{"clusterinfo"},
+		Run: func(cmd *cobra.Command, args []string) {
+			err := validateExactArgNumber(args, []string{"clusterinfo-file"})
+			kubeadmutil.CheckErr(err)
+
+			client, err := kubeconfigutil.ClientSetFromFile(*kubeConfigFile)
+			kubeadmutil.CheckErr(err)
+
+			// Here it's safe to get args[0], since we've validated that the argument exists above in validateExactArgNumber
+			clusterInfoFile := args[0]
```
Comment (jbeda) [inline cmd/kubeadm/app/cmd/phases/bootstraptoken.go]: Should we verify that this file doesn't have creds in it and meets other expectations?

**Review by jbeda — APPROVED** [review]
Forgot to review commands too.

File: cmd/kubeadm/app/cmd/init.go (hunk 6 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: @jbeda Yeah, we should probably some time look over style but not now :)
I agree with @mattmoyer, I personally prefer that style to hold the scope minimal for vars.

File: cmd/kubeadm/app/phases/addons/addons.go (hunk 14 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/addons.go]: This was because I centralized the "create-or-update-configmap" functionality which was used in two places; here and the cluster info code. I thought it would make sense to adopt the new func in both places... but no strong opinion here

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 16 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: We discussed this with @liggitt and @mikedanese in https://github.com/kubernetes/kubernetes/pull/49284
(You were mentioned on that PR as well :wink:)

File: cmd/kubeadm/app/cmd/phases/bootstraptoken.go (hunk 20 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/bootstraptoken.go]: That is not required as kubeadm just strips out and unnamed or currently used cluster from that file: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/kubeconfig/kubeconfig.go#L103

So it will never send creds. Not sure if we have to restrict having no creds here...

File: cmd/kubeadm/app/cmd/phases/phase_test.go (hunk 17 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/phase_test.go]: good catch! Copy-paste error :)

File: cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go (hunk 15 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/node/tlsbootstrap.go]: I think we should treat this as an implementation detail; we're not printing for each resource created right now; instead at each logical step. Having decent loglevel functionality would make a lot of sense here, but we don't have that rn

File: cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go (hunk 13 above)
Comment (luxas) [inline cmd/kubeadm/app/phases/bootstraptoken/clusterinfo/clusterinfo_test.go]: Just commit weirdness. It changed (wrongly) in the first commit and then reverted in the second.
Didn't squash properly it seems. Now fixed.

**Review by mattmoyer — APPROVED** [review]

**Comment by mattmoyer** [comment]
- `pull-kubernetes-unit` flake is https://github.com/kubernetes/kubernetes/issues/50262
 - `pull-kubernetes-verify` is failing because you dropped `cmd/kubeadm/app/util/token` from `hack/.golint_failures`

File: cmd/kubeadm/app/cmd/init.go (hunk 12 above)
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: We do name the phases, so +1.

Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Yeah, can be done in followups

**Comment by luxas** [comment]
Trivial rebase upon #50320, reapplying the label
